presto-hive/src/main/java/io/prestosql/plugin/hive/util/FieldSetterFactory.java

```diff
@@ -374,60 +368,31 @@ public void setField(Block block, int position)
         {
             super(rowInspector, row, field);
+            this.type = requireNonNull(type, "type is null");
             this.timeZone = requireNonNull(timeZone, "timeZone is null");
         }
 
-        protected Timestamp getTimestamp(long micros, int picosOfMicro)
-        {
-            long epochMillis = floorDiv(micros, MICROSECONDS_PER_MILLISECOND);
-            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
-            int nanosOfSecondFromMicros = floorMod(micros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
-            int nanosFromPicos = roundDiv(picosOfMicro, PICOSECONDS_PER_NANOSECOND);
-            // note that the nanos argument represents the nanos of second fraction, not nanos of millis, as one might expect
-            return Timestamp.ofEpochMilli(epochMillis, nanosOfSecondFromMicros + nanosFromPicos);
-        }
-    }
-
-    private static class ShortTimestampFieldSetter
-            extends TimestampFieldSetter
-    {
-        public ShortTimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, DateTimeZone timeZone)
-        {
-            super(rowInspector, row, field, timeZone);
-        }
-
         @Override
         public void setField(Block block, int position)
         {
-            long micros = TIMESTAMP_MICROS.getLong(block, position);
-            Timestamp timestamp = getTimestamp(micros, 0);
-            value.set(timestamp);
-            rowInspector.setStructFieldData(row, field, value);
-        }
-    }
-
-    private static class LongTimestampFieldSetter
-            extends TimestampFieldSetter
-    {
-        public LongTimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, DateTimeZone timeZone)
-        {
-            super(rowInspector, row, field, timeZone);
-        }
-
-        @Override
-        public void setField(Block block, int position)
-        {
-            LongTimestamp longTimestamp = (LongTimestamp) TIMESTAMP_NANOS.getObject(block, position);
-            Timestamp timestamp = getTimestamp(longTimestamp.getEpochMicros(), longTimestamp.getPicosOfMicro());
+            LongTimestamp longTimestamp = longTimestampRepresentation(block, position, (TimestampType) type);
+            long microsOfEpoch = longTimestamp.getEpochMicros();
+            long epochMillis = floorDiv(microsOfEpoch, MICROSECONDS_PER_MILLISECOND);
+            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
+            int nanosOfSecondFromMicros = floorMod(microsOfEpoch, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
+            int nanosFromPicos = roundDiv(longTimestamp.getPicosOfMicro(), PICOSECONDS_PER_NANOSECOND);
+            // note that the nanos argument represents the nanos of second fraction, not nanos of millis, as one might expect
+            Timestamp timestamp = Timestamp.ofEpochMilli(epochMillis, nanosOfSecondFromMicros + nanosFromPicos);
```

Comment:
Since the millis part is passed twice (😱  i know), we actually assume that `timeZone.convertLocalToUTC(epochMillis, false)` will not change the miilis part of the epochMillis (reasonable assumption). So we can simplify the code greatly by replacing `epochMillis` with `epochSeconds`.

Maybe sth along

```java
        @Override
        public void setField(Block block, int position)
        {
            LongTimestamp longTimestamp = longTimestampRepresentation(block, position, (TimestampType) type);

            long epochSeconds = floorDiv(longTimestamp.getEpochMicros(), MICROSECONDS_PER_SECOND);
            long picosOfSecond = (long) floorMod(longTimestamp.getEpochMicros(), MICROSECONDS_PER_SECOND) * PICOSECONDS_PER_MICROSECOND +
                    longTimestamp.getPicosOfMicro();

            epochSeconds = convertLocalEpochSecondsToUtc(epochSeconds);

            int nanosOfSecond = toIntExact(roundDiv(picosOfSecond, PICOSECONDS_PER_NANOSECOND));
            if (nanosOfSecond == NANOSECONDS_PER_SECOND) {
                // TODO add a test case for second overflow when rounding picos to nanos
                epochSeconds++;
                nanosOfSecond = 0;
            }

            Timestamp timestamp = Timestamp.ofEpochMilli(epochSeconds * MILLISECONDS_PER_SECOND, nanosOfSecond);
            value.set(timestamp);
            rowInspector.setStructFieldData(row, field, value);
        }

        private long convertLocalEpochSecondsToUtc(long epochSeconds)
        {
            long epochMillis = epochSeconds * MILLISECONDS_PER_SECOND;
            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
            verify(epochMillis % MILLISECONDS_PER_SECOND == 0, "Invalid epochMillis after converting epochSeconds %s with zone %s: %s", epochSeconds, timeZone, epochMillis);
            return epochMillis / MILLISECONDS_PER_SECOND;
        }
```